### ISSUE TITLE
Whitelist  wont work.

### ISSUE THREAD

**Issue by victordk** [jacobsalmela/pi-hole#152]
Hi

I have added several domains to whitelist.txt. Using example.com as syntax. The domains are plain http. 
A Gravity.sh run show they are added as whitelistet domains.

When accessing the domains dnsmasq ignores the added entries. (Taillogging shows the domains getting redirected to the pi.)

I have tried to restart the pi but the problem persists

Edit: The blacklistet domain did work after a few minutes.

**Comment by jacobsalmela** [comment]
The whitelist [isn't perfect yet](https://github.com/jacobsalmela/pi-hole/issues/147).  However, you can use `whitelist.sh`, which is much better and runs independently of `gravity.sh`.  The blacklist just appends entries, so there are no problems there.  

Also, you may need to whitelist more than just the site you want to visit.  See [this FAQ](http://pi-hole.net/faq/how-do-i-whitelist-or-blacklist-a-webiste-or-domain/) for more information.

**Comment by victordk** [comment]
Well i have whitelisted the subdomains. Used the source viewer in chrome, and the specific subdomains i have whitelisted still shows the blank html page from lighttpd. The site in question are nfl.com. Tried with whitelist.sh and adding the domains as argument. 

/usr/local/bin/whitelist.sh adobetag.com googletagservices.com www.adobetag.com www.googletagservices.com r.turn.com yui.yahooapis.com nfl.demdex.net imiclk.com combine.nflcdn.com

Gets the reply: ** 9 domain(s) whitelisted.

But they are still redirected.

**Comment by victordk** [comment]
And now on the second try - it works! Seems a tad flaky - i guess.

**Comment by jacobsalmela** [comment]
You can try restarting `dnsmasq` to make sure your cache is completely flushed.  Also make sure the browser cache is flushed, too.

**Comment by shadowkhas** [comment]
Apologies if it's bad GitHub etiquette to comment on a previously-closed issue, but it's identical to what's happening to me, too - I'm having trouble with whitelist.sh as well. Funnily enough, I'm trying to whitelist yui.yahooapis.com as well, and it doesn't seem to carry though to client devices.

I'll run "whitelist.sh yui.yahooapis.com", and I get a confirmation message of a domain whitelisted.
When I run "dig yui.yahooapis.com", I receive proper DNS responses, not the PiHole address. However, using any client devices (or curl on the PiHole device itself) returns the the redirect .html that I have set up.
I've tried restarting dnsmasq using "sudo service dnsmasq stop" and "sudo service dnsmasq start", and also "sudo /etc/init.d/dnsmasq stop" followed by "sudo /etc/init.d/dnsmasq start" and the issue persists, through a restart as well. Hope that can maybe shed some extra light.

**Comment by dschaper** [comment]
What are your client machines? Windows machines will cache the response, in addition to restarting dnsmasq you'll need to flush the dns cache on Windows machines or wait until the timeout period passes.

**Comment by jacobsalmela** [comment]
For that particular domain, I can reproduce the problem even after removing both IPv4 and IPv6 entries from `/etc/pihole/gravity.list`.  I'll re-open and classify as a bug.

**Comment by shadowkhas** [comment]
Thanks for the quick responses! Happy to be of any further help if needed, I've got Windows, OS X, and the Pi itself to test with if it'll be of any help down the line.

**Comment by PromoFaux** [comment]
Hi there, not sure if this is worth a new issue...

I also am having issues with the whitelist. I cannot for the life of me seem to whitelist google-analytics.com.

Looking at the following lines in gravity.sh:

      # Remove whitelist entries from list
        grep -vxf $latentWhitelist $piholeDir/$matter > $piholeDir/$andLight

There are still multiple entires for google-analytics.com in the andlight file after this is run. Is this expected behaviour? My understanding was that the andlight file contains everything from the matter file except what is on the whitelist...

Pulling my hair out here!

**Comment by dschaper** [comment]
Are you using the `whitelist.sh` script to add the entry, or are you editing the `whitelist.txt` file? It seems that the script is more effective in whitelisting.

**Comment by jacobsalmela** [comment]
@PromoFaux, we already have two whitelist issues classified as bugs, so no need for another.

You will probably have better luck with the `whitelist.sh` script.  The lines you mentioned above will probably be replaced with [something like this](https://github.com/jacobsalmela/pi-hole/issues/147#issuecomment-169216344), but with the ability to remove both IPv4 and IPv6 entries.

I have noticed that the function inside `gravity.sh` really doesn't work as well.  I think the logic is just a bit off.

If you have been editing `whitelist.txt`, you should be able to still apply it with this command (I haven't tested though):
```
/usr/local/bin/whitelist.sh $(cat /etc/pihole/whitelist.txt)
```

**Comment by ajp2k13** [comment]
Maybe you guys can get some ideas from this thread? I ran it on my Asus router before I got the Pi-Hole..
http://www.snbforums.com/threads/adblocking-with-combined-hosts-file.15309/

**Comment by PromoFaux** [comment]
@dschaper I've tried both ways using fresh installs, so far the only effective solution i've found is to remove the entries from the gravity.list file.

@jacobsalmela Apologies, I just read issue #147, and realised it's more or less the same bug that I was describing before. I must have missed it :)

>If you have been editing whitelist.txt, you should be able to still apply it with this command (I haven't tested though):

>    /usr/local/bin/whitelist.sh $(cat /etc/pihole/whitelist.txt)

Did you mean gravity.sh? I assume that just goes in place of the line that adds whitelist.txt to latentwhitelist.txt?

**Comment by jacobsalmela** [comment]
fixed by #174
